docker-build.rb

```diff
@@ -0,0 +1,79 @@
+require "fileutils"
+require 'optparse'
+
+image_name = "router"
+
+def exec_or_die( cmd, current_dir = nil )
+	orgDir = Dir.pwd
+	Dir.chdir(current_dir) if current_dir != nil
+	
+	res = `#{cmd}`
+	
+	Dir.chdir(orgDir)
+
+	if $?.exitstatus != 0
+		print(res)
+		exit(false)
+	end
+end
+
+options = {}
+OptionParser.new do |opts|
+  opts.banner = "Usage: docker-build.rb [options]"
+  opts.on("-t", "--dockertag TAG", "Docker tag to use when pusing the image. Defaults to latest.") do |dt|
+    options[:dockertag] = dt
+  end
+  opts.on("-r", "--dockerregistry REGISTRY", "The Docker registry to use. Defaults to Docker hub (imqs namesapce).") do |dr|
+    options[:dockerregistry] = dr
+  end
+  opts.on("-u", "--dockeruser USER", "The Docker user to login with.") do |du|
+    options[:dockeruser] = du
+  end
+  opts.on("-p", "--dockerpass PASS", "The Docker password to login with.") do |dp|
+    options[:dockerpass] = dp
+  end  
+  opts.on("-h", "--help", "Prints this help") do
+	puts opts
+    exit
+  end
+end.parse!
+
+puts("Building router binary")
+exec_or_die("docker run --rm -e GOPATH=/usr/src/router -v #{Dir.pwd}:/usr/src/router -w /usr/src/router golang:1.8 go install -ldflags \"-linkmode external -extldflags -static\" github.com/IMQS/router-core
+")
+puts("Building image")
+exec_or_die("docker build -t router:#{options[:dockertag]} .")
+
+tag_command = nil
```

Comment:
I think this can be quite a bit simpler:

1. Don't allow any optional parameters to docker-build.rb. Instead, require only one mandatory parameter, which is the git branch that is being built. The Jenkins invocation then becomes `ruby docker-build.rb $GIT_BRANCH`
2. Don't login. Rather login to docker hub via a cronjob on the CI slave
3. Don't allow arbitrary registries

These images are so small (like 10 MB), that I don't think it's worth it pushing them to an internal registry, ever.

I've come to believe that if your docker images are large with every build, then you're not layering them enough. In this case, that's not even an issue, but for a larger build, say a Java project, the JDK portion remains fairly constant. So your new docker build should only contain the little bit of code that we actually write. Your container should be based on an image that changes infrequently.

Reply:
Noted. This was just to give us some flexibility if we needed to use alternate registries. We can simplify it if we make that call.